src/vanilla/store.ts

```diff
@@ -488,36 +488,44 @@ const buildStore = (getAtomState: StoreArgs[0]): Store => {
     atomState: AtomState<Value>,
     ...args: Args
   ): Result => {
-    const getter: Getter = <V>(a: Atom<V>) =>
-      returnAtomValue(readAtomState(pending, a, getAtomState(a, atomState)))
-    const setter: Setter = <V, As extends unknown[], R>(
-      a: WritableAtom<V, As, R>,
-      ...args: As
-    ) => {
-      const aState = getAtomState(a, atomState)
-      let r: R | undefined
-      if (isSelfAtom(atom, a)) {
-        if (!hasInitialValue(a)) {
-          // NOTE technically possible but restricted as it may cause bugs
-          throw new Error('atom not writable')
+    let isSync = true
+    try {
+      const getter: Getter = <V>(a: Atom<V>) =>
+        returnAtomValue(readAtomState(pending, a, getAtomState(a, atomState)))
+      const setter: Setter = <V, As extends unknown[], R>(
+        a: WritableAtom<V, As, R>,
+        ...args: As
+      ) => {
+        const aState = getAtomState(a, atomState)
+        let r: R | undefined
+        if (isSelfAtom(atom, a)) {
+          if (!hasInitialValue(a)) {
+            // NOTE technically possible but restricted as it may cause bugs
+            throw new Error('atom not writable')
+          }
+          const hasPrevValue = 'v' in aState
+          const prevValue = aState.v
+          const v = args[0] as V
+          setAtomStateValueOrPromise(a, aState, v)
+          mountDependencies(pending, a, aState)
+          if (!hasPrevValue || !Object.is(prevValue, aState.v)) {
+            addPendingAtom(pending, a, aState)
+            recomputeDependents(pending, a, aState)
+          }
+        } else {
+          r = writeAtomState(pending, a, aState, ...args) as R
         }
-        const hasPrevValue = 'v' in aState
-        const prevValue = aState.v
-        const v = args[0] as V
-        setAtomStateValueOrPromise(a, aState, v)
-        mountDependencies(pending, a, aState)
-        if (!hasPrevValue || !Object.is(prevValue, aState.v)) {
-          addPendingAtom(pending, a, aState)
-          recomputeDependents(pending, a, aState)
+        if (!isSync) {
+          flushPending(pending)
         }
```

Comment:
Edge case: do we care if this changes the behavior a little bit for subscriber listeners?

```ts
const a = atom(0)
const w = atom(null, (get, set) => {
  console.log('before write')
  set(a, v=> v + 1)
  console.log('after write')
})
w.onMount = (setSelf) => {
  setSelf()
  console.log('after setSelf')
}
store.sub(a, () => {
  console.log('a listener fired')
})
store.sub(w, () => {})
/*
  CURRENT:
    1. 'before write'
    2. 'a listener fired'
    3. 'after write'
    4. 'after setSelf'

  NEW:
    1. 'before write'
    2. 'after write'
    3. 'after setSelf'
    4. 'a listener fired'
*/ 
```

Reply:
Is it why the current test is failing?

Do you know how to make a failing test if we don't have `isSync`?

Reply:
> Is it why the current test is failing?

Yes

> Do you know how to make a failing test if we don't have `isSync`?

Thinking on this more...
I think we need to keep this synchronous call to `flushPending` in `set`. Some third-party libraries depend on this behavior, so we need some way to call all subscribers.

Reply:
Yes, I agree. This is better.

Hmm, it looks like atomWithObservable and atomWithStorage are not failing after all.